reps/2023-02-03-node-affinity-feature-enhancements.md

```diff
@@ -0,0 +1,367 @@
+    ).remote()
+
+```
+
+**2. Anti-Affinity to the nodes of the special node_id.
+```
+actor_1 = Actor.options(
+        scheduling_strategy=node_affinity(label_not_in(key="node_id", values=["xxxxx", "aaaaa"], is_soft=false))
+    ).remote()
+```
+
+**3. Anti-Affinity to the node which have GPU.
+```
+actor_1 = Actor.options(
+        scheduling_strategy=node_affinity(label_does_not_exist(key="gpu_type", is_soft=false))
+    ).remote()
+```
+
+**4. It must be scheduled to a node whose region is china, and as much as possible to a node whose availability zone is china-east-1. **
+
+```
+actor_1 = Actor.options(
+        scheduling_strategy=node_affinity([
+            label_in("region", ["china"], is_soft=false),
+            label_in("zone", ["china-east-1"], is_soft=true)
+        ])
+    ).remote()
+```
+**5. NodeAffinity can be used together with the CustomResource mechanism.
+These two mechanisms are completely non-conflicting.  
+```
+actor_1 = Actor.options(
+                num_gpus=1,
+                resources={"custom_resource": 1},
+                scheduling_strategy=node_affinity([
+                    label_exist("gpu_type", is_soft=false),
+                    label_in("gpu_type", ["A100", "P100"], is_soft=true),
+                ]).remote()
+```
+
+FAQ:
+This node static labels seems to be very similar to custom_resources. Why not use custom_resources to instance of node_affinity?
+
+1. The semantics of static labels and custom_resources are completely different. custom_resources is a countable resource for nodes. And labels is an attribute or identifier of a node.
+2. Now custom_resources cannot implement the function of anti-affinity.
+3. Using labels to classify nodes and implement affinity/anti-affinity will be more natural and easier for users to accept.
+
+### Implementation plan
+
+![LabelsAffinity](https://user-images.githubusercontent.com/11072802/203686851-894536d0-86fa-4fab-a5fe-2e656ac198e5.png)
+
+1. Add the labels data structure to the resource synchronization data structure(NodeResources). 
+```
+NodeResources {
+  ResourceRequest total;
+  ResourceRequest available;
+  /// Only used by light resource report.
+  ResourceRequest load;
+  /// Resources owned by normal tasks.
+  ResourceRequest normal_task_resources
+  /// Map<label_type, Map<namespace, Map<label_key, label_value>>> Nodes labels information
+  absl::flat_hash_map<string, absl::flat_hash_map<string, absl::flat_hash_map<string, string>>> labels;
+}
+```
+
+2. When a node registers with GCS and broadcasts to all nodes, add the static labels information to NodeResources.
+
+3. Scheduling optimization through Labels  
+Now any node raylet has node static labels information for all nodes.  
+when NodeAffinity schedules, if it traverses the Labels of each node, the algorithm complexity is very large, and the performance will be poor.   
+** Therefore, it is necessary to generate a full-cluster node labels index table to improve scheduling performance. </b>
```

Comment:
What's the index strategy for the more complex expressions such as negation and containment? I wonder if we need to cache the list of valid nodes per task/expression to avoid O(n^2) scheduling slowdowns with many nodes and tasks.

Reply:
This tag indexing strategy is actually very simple, I added it in the REP. The tag index only needs to get the nodes with these key&values.
![image](https://user-images.githubusercontent.com/11072802/217146026-762c5cb0-b608-46bd-88c8-7ece2bd9c2ee.png)